addons/mass_mailing/tests/test_mailing_internals.py

```diff
@@ -474,6 +475,98 @@ def test_mailing_shortener(self):
                     link_params=link_params,
                 )
 
+    @users('user_marketing')
+    @mute_logger('odoo.addons.mail.models.mail_mail')
+    def test_mailing_launch_flow(self):
+        """ Test that mail should be sent directly if total recipients are less than value
+        specified by config param 'mass_mailing_immediate_send_nbr', otherwise it should
+        enqueue the mailing """
+
+        # number of emails are less then specified limit
+        mailing_1 = self.env['mailing.mailing'].create({
+            'name': 'One',
+            'subject': 'One',
+            'mailing_model_name': 'mailing.list',
+            'contact_list_ids': [(4, self.mailing_list_1.id)],
+            'body_html': 'This is mass mail marketing demo',
+        })
+
+        # number of emails are more then specified limit
+        mailing_2 = self.env['mailing.mailing'].create({
+            'name': 'Two',
+            'subject': 'Two',
+            'mailing_model_name': 'mailing.list',
+            'contact_list_ids': [(4, self.mailing_list_2.id)],
+            'body_html': 'This is mass mail marketing demo',
+        })
+        with self.mock_mail_gateway(mail_unlink_sent=False), patch.object(MassMailing, '_get_mass_mailing_immediate_send_nbr', return_value=3):
+            (mailing_1 | mailing_2).action_launch()
+        self.assertEqual(mailing_1.state, 'done')
+        self.assertEqual(mailing_2.state, 'in_queue')
+        self.assertEqual(len(self._new_mails.exists()), 3)
+        self.assertEqual(len(self._mails), 3)
+        # test that mailing_1 mails are sent only
+        for mail in self._mails:
+            self.assertEqual(mail['subject'], 'One')
+
+        # enqueue emails if specified limit is set to 0
+        mailing_3 = self.env['mailing.mailing'].create({
+            'name': 'Two',
+            'subject': 'Two',
+            'mailing_model_name': 'mailing.list',
+            'contact_list_ids': [(4, self.mailing_list_1.id), (4, self.mailing_list_2.id)],
+            'body_html': 'This is mass mail marketing demo',
+        })
+        with self.mock_mail_gateway(mail_unlink_sent=False), patch.object(MassMailing, '_get_mass_mailing_immediate_send_nbr', return_value=0):
+            mailing_3.action_launch()
+        self.assertEqual(mailing_3.state, 'in_queue')
+        self.assertEqual(len(self._new_mails.exists()), 0)
```

Comment:
you can use `assertNotSentEmail`

Reply:
I have already used assertNotSentEmail in next line. For double confirmation I have done this. If you insist I can remove this.